### PR TITLE
1490 sqlite term server

### PR DESCRIPTION
**Pull request by jonahkaye**
Ticket: #[1490](https://github.com/metriport/metriport-internal/issues/1490)

### Description

- embedded term server typescript w/ sqlite
- I tried to make the term server as similar to the fhir converter as possible. No dependencies, simple infra deploy creating docker from the package
- In order for the server to run with the fully loaded ready to go term server, it needs to download the database from s3 `metriport-multi-purpose ` bucket. It was either storing the 4gig compressed terminology itself on s3 and having cicd take longer, or just storing the smaller 500MB db on s3. 
- for understanding structure of term server [read](https://www.notion.so/metriport/UMLS-Metathesaurus-30719d9bf0984db6b2795696163244e8?pvs=26&qid=) and watch [loom](https://www.loom.com/share/1325784c31de479b840d3a29bcfb5b42) 

### Todo

- [x] Crosswalk
- [x] infra - ecs or lambda (could load db file from s3)
- [x] CICD
- [x] fhir validation test on a Condition with multiple icd10 codes
- [x] how to handle medplum license - attaching reference to apache license on explicitly copied files

### Testing

- Local
  - [x] local import
- Staging
  - [x] branch to staging
  - [ ] staging

### Artillery Load Test

```
--------------------------------
Summary report @ 19:46:23(-0700)
--------------------------------

vusers.completed: .............................................................. 100000
vusers.created: ................................................................ 100000
vusers.created_by_name.lookupCode: ............................................. 49937
vusers.created_by_name.translateCode: .......................................... 50063
vusers.failed: ................................................................. 0
vusers.session_length:
  min: ......................................................................... 76
  max: ......................................................................... 765.1
  mean: ........................................................................ 101.1
  median: ...................................................................... 87.4
  p95: ......................................................................... 144
  p99: ......................................................................... 478.3
```

### Release Plan

- [ ] Before the deploy, add the terminology.db file to the multi-purpose bucket so the application doesnt bootloop
- [ ] Added to [monthly product update](https://www.notion.so/metriport/Customer-Updates-21b4e9d3ad5f4fd68db587a11db28cff?pvs=4) (introduce a feature that would be useful customers)
- [ ] Merge this

### CODE REVIEW

File: packages/terminology/package.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/metriport/metriport/issues"
+  },
+  "homepage": "https://github.com/metriport/metriport#readme",
+  "dependencies": {
+    "@medplum/core": "^3.2.5",
+    "@medplum/fhir-router": "^3.2.5",
+    "@medplum/fhirtypes": "^3.2.5",
```
Comment (jonahkaye) [inline packages/terminology/package.json:26]: note we need a higher version of medplum here than in the rest of the codebase

File: packages/terminology/src/seed/seedUmlsLookup.ts
Comment (jonahkaye) [inline packages/terminology/src/seed/seedUmlsLookup.ts:1]: mostly from [here](https://github.com/medplum/medplum/pull/3553/files#diff-211ded7d59086782a6c0fe22eb0bfa3b5aeaf3ad90d916cdea69a08524b1641c)

File: packages/terminology/src/sqlite.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    if (migration) {
+      console.log("Running migration", migration.sql);
+      await dbClient.run(migration.sql);
+      await dbClient.run(`UPDATE "DatabaseMigration" SET "version"=${i}`);
+    }
+  }
+}
+
+// Function to initialize SQLite
+export async function initSqliteFhirServer(): Promise<void> {
```
Comment (jonahkaye) [inline packages/terminology/src/sqlite.ts:42]: init flow:
1. load DB
2. perform migrations
3. seed DB with code systems which are prerequisite to import, lookup, translate operations

File: packages/terminology/src/sqlClient.ts
Comment (jonahkaye) [inline packages/terminology/src/sqlClient.ts:1]: sqlite basic operations are not async. Need to wrap in promise to make async and awaitable

File: packages/terminology/src/sqlite.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import * as migrations from "./migrations/index";
+import { seedCodeSystems } from "./seed";
+
+let dbClient: SqliteClient | undefined;
+
+type Migration = {
+  sql: string;
+};
+
+async function migrate(dbClient: SqliteClient): Promise<void> {
```
Comment (jonahkaye) [inline packages/terminology/src/sqlite.ts:14]: migration table stores current version. 
If version file v5.ts exists, but we are only on v4, then we run the sql query in v5.ts

File: packages/terminology/src/operations/codeLookup.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/codeLookup.ts:1]: mostly copied from here: https://github.com/medplum/medplum/blob/32b1f646efb32297736a4481024553c1ec14bea2/packages/server/src/fhir/operations/codesystemlookup.ts#L16

File: packages/terminology/src/operations/codeImport.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/codeImport.ts:1]: mostly from [here](https://github.com/medplum/medplum/blob/32b1f646efb32297736a4481024553c1ec14bea2/packages/server/src/fhir/operations/codesystemimport.ts)

File: packages/terminology/src/operations/definitions/codeSystem.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/definitions/codeSystem.ts:1]: used for seeding

File: packages/terminology/src/sqlite.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    console.error("Error creating database directory:", err);
+    throw err;
+  }
+
+  dbClient = new SqliteClient(dbPath);
+  await migrate(dbClient);
+  await seedCodeSystems(dbClient);
+}
+
+export function getSqliteClient(): SqliteClient {
```
Comment (jonahkaye) [inline packages/terminology/src/sqlite.ts:59]: getter for accessing the sclite client

File: packages/terminology/src/operations/utils/parameters.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/utils/parameters.ts:1]: from [here](https://github.com/medplum/medplum/blob/32b1f646efb32297736a4481024553c1ec14bea2/packages/server/src/fhir/operations/utils/parameters.ts)

used for contract formatting to keep things in the formats Medplum has for requests and responses

File: packages/terminology/package-lock.json
Comment (jonahkaye) [inline packages/terminology/package-lock.json:1]: is it weird that we have a package-lock.json for this? 
is it because I introduce different versions for medplum than for the rest of the codebase

File: packages/utils/src/terminology/index.ts
Comment (jonahkaye) [inline packages/utils/src/terminology/index.ts:1]: simple util for hitting term server

File: packages/terminology/src/migrations/v1.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export const sql = `
+    CREATE TABLE IF NOT EXISTS CodeSystem (
+      id UUID NOT NULL PRIMARY KEY,
+      content TEXT NOT NULL,
```
Comment (jonahkaye) [inline packages/terminology/src/migrations/v1.ts:4]: i think SQLITE actually recently added JSONB support. 
would be better to use that then stringify(text)

File: packages/terminology/src/sqlClient.ts
Comment (leite08) [reply packages/terminology/src/sqlClient.ts:1]: Can probably use [Node's promisify](https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original) to avoid writing all that plumbing code. 😃

File: packages/terminology/package-lock.json
Comment (leite08) [reply packages/terminology/package-lock.json:1]: In order to be part of npm's workspace, we need to add `packages/terminology` to the root `package.json`

File: packages/terminology/src/router.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  const fhirRequest = createFhirRequest(req);
+  const response = await codeSystemImportHandler(fhirRequest);
+  res.status(200).json(response);
+});
+
+fhirRouter.post("/CodeSystem/lookup", async (req: Request, res: Response) => {
+  const fhirRequest = createFhirRequest(req);
+  const response = await codeSystemLookupHandler(fhirRequest);
+  res.status(200).json(response);
+});
```
Comment (leite08) [inline packages/terminology/src/router.ts:29]: Any reason not to use the approach/code we have on the OSS API?

File: packages/terminology/src/operations/codeImport.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    const codeSystem = await findCodeSystemResource(params.system);
+    await importCodeSystemSqlite(codeSystem, params.concept, params.property);
+    return [allOk, buildOutputParameters(operation, codeSystem)];
+  } catch (err) {
+    return [normalizeOperationOutcome(err)];
+  }
+}
+
+export async function importCodeSystemSqlite(
```
Comment (leite08) [inline packages/terminology/src/operations/codeImport.ts:64]: I suggest we don't name our stuff w/ tech-specific lingo.

File: packages/terminology/src/sqlClient.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import { Database } from "sqlite3";
+/* eslint-disable */
+
+export class SqliteClient {
```
Comment (leite08) [inline packages/terminology/src/sqlClient.ts:4]: Since we're building this, I'd have an interface `DbClient` and a class `SqliteDbClient` that implements that interface, with a factory that for now just returns a new instance of the class.

File: packages/terminology/src/config.ts
Comment (jonahkaye) [inline packages/terminology/src/config.ts:1]: had to add this so term server would have no dependency on shared/core

File: packages/terminology/src/operations/conceptMapImport.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
+import { FhirRequest, FhirResponse } from "@medplum/fhir-router";
+import { ConceptMap } from "@medplum/fhirtypes";
+import { normalizeOperationOutcome } from "@medplum/core";
+import { v4 as uuidv4 } from "uuid";
+
+import { getSqliteClient } from "../sqlite";
+
+function generateReverseConceptMaps(conceptMap: ConceptMap): ConceptMap[] {
```
Comment (jonahkaye) [inline packages/terminology/src/operations/conceptMapImport.ts:8]: the idea here is to create bidirectional concept maps. 
If we know a => b, then why not also put b => a into the term server too.

File: packages/terminology/src/operations/conceptMapImport.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        },
+      ],
+    };
+    reverseMaps.push(reverseMap);
+  }
+
+  return reverseMaps;
+}
+
+export async function conceptMapImportHandler(
```
Comment (jonahkaye) [inline packages/terminology/src/operations/conceptMapImport.ts:47]: there are a lot of codes with multiple entries. We need to do upserts here to support that.

File: packages/terminology/src/operations/definitions/codeLookupOperation.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/definitions/codeLookupOperation.ts:1]: copied

File: packages/terminology/src/operations/definitions/codeSystem.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/definitions/codeSystem.ts:1]: copied

File: packages/terminology/src/operations/definitions/conceptMapTranslate.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/definitions/conceptMapTranslate.ts:1]: copied

File: packages/terminology/src/operations/utils/parameters.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/utils/parameters.ts:1]: copied

File: packages/terminology/src/app.ts
Comment (jonahkaye) [inline packages/terminology/src/app.ts:1]: from scratch

File: packages/terminology/src/migrations/index.ts
Comment (jonahkaye) [inline packages/terminology/src/migrations/index.ts:1]: rewrote these DBs to be sqlite

File: packages/terminology/src/operations/codeImport.ts
Comment (jonahkaye) [reply packages/terminology/src/operations/codeImport.ts:1]: just rewrote logic to use sqlite

File: packages/terminology/src/operations/codeLookup.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/codeLookup.ts:1]: Mostly copied. just rewrote logic to use sqlite

File: packages/terminology/src/operations/conceptMapImport.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/conceptMapImport.ts:1]: wrote from scratch. Wasnt in medplum

File: packages/terminology/src/operations/conceptMapTranslate.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/conceptMapTranslate.ts:1]: copied + sqlite

File: packages/terminology/src/operations/utils/codeSystemLookup.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/utils/codeSystemLookup.ts:1]: copied + sqlite

File: packages/terminology/src/router.ts (hunk 6 above)
Comment (jonahkaye) [reply packages/terminology/src/router.ts:31]: i copied this mostly

File: packages/terminology/src/s3.ts
Comment (jonahkaye) [inline packages/terminology/src/s3.ts:1]: had to copy this from core so term server wouldnt have core dependency

File: packages/infra/lib/api-stack/terminology-server-service.ts [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+  const bucketName = "umls-terminology-db";
+  // const terminologyBucket = new s3.Bucket(stack, "TerminologyBucket", {
+  //   bucketName,
+  //   publicReadAccess: false,
+  //   encryption: s3.BucketEncryption.S3_MANAGED,
+  //   versioned: true,
+  // });
+
+  const terminologyBucket = s3.Bucket.fromBucketName(stack, "TerminologyBucket", bucketName);
```
Comment (jonahkaye) [inline packages/infra/lib/api-stack/terminology-server-service.ts:59]: before PR is approved, I can change this back to be a create Bucket instead of a from Bucket. Used from since it was easier for dev and I didnt have to deal with permissions to delete the DB before each branch to staging to stop resource conflicts

File: .github/workflows/deploy-production.yml
Comment (jonahkaye) [inline .github/workflows/deploy-production.yml:1]: cicd here modelled after fhir-converter. Similar standalone packages with 0 dependencies on the rest of metriport

File: LICENSE-APACHE
Comment (jonahkaye) [inline LICENSE-APACHE:1]: license for select files fully copied from medplum

File: packages/terminology/src/operations/definitions/codeLookupOperation.ts [hunk 12]
```diff
@@ -0,0 +1,302 @@
+// Licensed under Apache. See LICENSE-APACHE in the repo root for license information.
```
Comment (jonahkaye) [inline packages/terminology/src/operations/definitions/codeLookupOperation.ts:1]: apache

File: packages/terminology/src/seed/seedUmlsCrosswalk.ts [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+      {
+        source: config.system,
+        target: config.target,
+        element: [
+          {
+            code: concept.FROMID,
+            target: [
+              {
+                code: concept.TOID,
+                //TODO could perform a lookup to get display name or integrate with the rest of the concept flow
```
Comment (jonahkaye) [inline packages/terminology/src/seed/seedUmlsCrosswalk.ts:163]: seems important for application use cases.

**Review by RamilGaripov — COMMENTED** [review]
Just a partial review for now. Will pick up tmrw morning

File: packages/terminology/src/operations/conceptMapTranslate.ts
Comment (jonahkaye) [inline packages/terminology/src/operations/conceptMapTranslate.ts:1]: mostly copied with some simplifications and changes to use Sqlite

File: packages/terminology/src/migrations/v1.ts [hunk 14]
```diff
@@ -0,0 +1,6 @@
+export const sql = `
+    CREATE TABLE IF NOT EXISTS CodeSystem (
```
[the remark this answers is not in the document]
Comment (Goncharo) [reply packages/terminology/src/migrations/v1.ts:2]: +1 for the table names

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: packages/utils/src/terminology/load-test.yml [hunk 15]
```diff
@@ -0,0 +1,24 @@
+config:
+  target: 'http://localhost:3000'
```
Comment (Goncharo) [inline packages/utils/src/terminology/load-test.yml:2]: ditto 3000

File: packages/terminology/src/router.ts (hunk 6 above)
Comment (leite08) [reply packages/terminology/src/router.ts:31]: This is not using the `asyncHandler`. Any particular reason? It does things!

File: packages/terminology/src/config.ts
Comment (leite08) [reply packages/terminology/src/config.ts:1]: Why no dependency on Shared/Core? At least Shared should be safe to import - even the Dash does it.

File: packages/terminology/src/sqlite.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+import { S3Utils } from "./s3";
+
+import { TermServerClient, createTermServerClient } from "./sqlClient";
+import * as migrations from "./migrations/index";
+import { seedCodeSystems } from "./seed/seedCodeSystem";
+import { Config } from "./config";
+
+const region = Config.getAWSRegion();
+
+let dbClient: TermServerClient | undefined;
```
Comment (leite08) [inline packages/terminology/src/sqlite.ts:13]: We should use the more abstract representation, so we can really leverage the fact we have an interface in case we want to switch implementations.

File: packages/terminology/src/sqlite.ts
Comment (leite08) [inline packages/terminology/src/sqlite.ts:1]: This file has things that are not related to SQLite. I'd rename it to something more generic.

**Review by leite08 — COMMENTED** [review]
Leaving some comments, can finish reviewing later

File: packages/infra/lib/api-stack/terminology-server-service.ts [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+
+  fargateService.service.connections.allowFrom(
+    ec2.Peer.ipv4(vpc.vpcCidrBlock),
+    ec2.Port.allTraffic(),
+    "Allow traffic from within the VPC to the service secure port"
+  );
+
+  fargateService.service.connections.allowFromAnyIpv4(ec2.Port.allTcp());
+
+  fargateService.targetGroup.setAttribute("deregistration_delay.timeout_seconds", "17");
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/infra/lib/api-stack/terminology-server-service.ts:131]: I copied the fhir-converter-service

File: packages/terminology/src/config.ts
Comment (jonahkaye) [reply packages/terminology/src/config.ts:1]: So I looked at api and core and they both import shared via the file system with `"@metriport/shared": "file:packages/shared",`
I guess I could do what the dash does though and use whatever version of shared is on npm.

File: packages/terminology/README.md [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+## Overview
+
+This project provides a local term server for lookups and crosswalks between SNOMED and ICD.
+
+## Download Metathesaurus
+
+You can download the Metathesaurus from the following link:  
+[Metathesaurus Download](https://www.nlm.nih.gov/research/umls/licensedcontent/umlsknowledgesources.html)  
+**Note:** The file is large—4 GB compressed and 28 GB uncompressed. Releases occur twice a year, so stay alert!
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/terminology/README.md:11]: https://github.com/metriport/metriport-internal/issues/2129

File: packages/terminology/src/app.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+import dayjs from "dayjs";
+import duration from "dayjs/plugin/duration";
+
+dayjs.extend(duration);
+
+async function main() {
+  const app = express();
+
+  app.use(express.json({ limit: "50mb" }));
+  await initTermServer();
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/terminology/src/app.ts:16]: Takes around 10s. It worked fine when branching to staging.

File: packages/terminology/src/router.ts (hunk 6 above)
Comment (jonahkaye) [reply packages/terminology/src/router.ts:31]: what does it do

File: packages/terminology/src/seed/seedUmlsCrosswalk.ts [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+      continue;
+    }
+
+    const key = concept.MAPSETCUI + "|" + concept.FROMID;
+
+    const sourceParameters = createLookupParameters(C5885096.system, concept.FROMID);
+    const sourceLookup = await client.lookupCode(sourceParameters);
+    const sourceDisplay = sourceLookup[0].display;
+
+    if (concept.TOID.endsWith("?")) {
```
Comment (jonahkaye) [inline packages/terminology/src/seed/seedUmlsCrosswalk.ts:257]: tldr here is that some icd10 codes end in a ? and those map to a range of different icd10 codes based on what the 7th character is. The 7th character is a kind of toggleable field in icd10 that indicates some attributes about the condition. So we need to lookup all the possible variation of the code and then map all of those

File: .github/workflows/deploy-staging.yml [hunk 21]
```diff
@@ -57,6 +59,11 @@ jobs:
               - "packages/fhir-converter/src/**"
               - "packages/fhir-converter/test/**"
               - "packages/fhir-converter/deploy/**"
+            terminology:
+              - "packages/fhir-converter/Dockerfile"
+              - "packages/fhir-converter/package*.json"
+              - "packages/fhir-converter/src/**"
```
Comment (leite08) [inline .github/workflows/deploy-staging.yml:65]: missed updating the folder here

File: .github/workflows/deploy-production.yml [hunk 22]
```diff
@@ -60,6 +62,11 @@ jobs:
               - "packages/fhir-converter/test/**"
               - "packages/fhir-converter/deploy/**"
               - "packages/infra/**"
+            terminology:
+              - "packages/fhir-converter/Dockerfile"
+              - "packages/fhir-converter/package*.json"
+              - "packages/fhir-converter/src/**"
```
Comment (leite08) [inline .github/workflows/deploy-production.yml:68]: ditto: update folder

File: packages/infra/lib/api-stack/terminology-server-service.ts (hunk 11 above)
Comment (leite08) [reply packages/infra/lib/api-stack/terminology-server-service.ts:59]: Do we expect to store lots of files for the term server? From what I can tell, no, right? Then I'd suggest using the `metriport-multi-purpose-<env>` one.

File: .github/workflows/deploy-staging.yml
Comment (leite08) [inline .github/workflows/deploy-staging.yml:1]: We're missing building/testing the term server, right?

File: .github/workflows/pull-request_term-server.yml [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+
+      - name: Install
+        run: |
+          pwd
+          npm install
+        working-directory: ${{env.working-directory}}
+
+      - name: Build (linters)
+        run: npm run build
+        working-directory: ${{env.working-directory}}
```
Comment (leite08) [inline .github/workflows/pull-request_term-server.yml:50]: Let's add a test step as well, even if the respective npm script just logs "no tests" for now, so the plumbing is in place, please.

File: LICENSE-APACHE
Comment (leite08) [inline LICENSE-APACHE:1]: Why on the root folder? This is related to the term server, right? So I think it should be on the respective folder as `LICENSE.md`.

File: packages/terminology/package.json
Comment (leite08) [inline packages/terminology/package.json:1]: We can make this private, as we're not publishing this to NPM, right?

File: packages/terminology/src/app.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (leite08) [reply packages/terminology/src/app.ts:16]: It seems pretty close. Check the `configureHealthCheck` on the infra, it seems it will try twice, with 10s between each check. If this takes around 10s, it could stretch to 20 and fail.

The solution might be just to update the healthchecks.

**Review by leite08 — COMMENTED** [review]
Are we not adding the term server as part of the npm workspace?

I'd expect to have it on the root `package.json` as well, so commands there would be replicated to the term service, like `build`, `test`, etc.

File: LICENSE-APACHE
Comment (jonahkaye) [reply LICENSE-APACHE:1]: @Goncharo and I went through this and root folder is where he suggested it go

File: packages/terminology/src/seed/seedUmlsCrosswalk.ts (hunk 13 above)
Comment (jonahkaye) [reply packages/terminology/src/seed/seedUmlsCrosswalk.ts]: this was completed

File: packages/terminology/src/app.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/terminology/src/app.ts:21]: I moved the health check to be after the term server init so this shouldnt be an issue

File: packages/terminology/src/router.ts [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+    headers: req.headers,
+  };
+}
+
+fhirRouter.post(
+  "/CodeSystem/import",
+  asyncHandler(async (req: Request, res: Response) => {
+    const fhirRequest = createFhirRequest(req);
+    const response = await codeSystemImportHandler(fhirRequest);
+    res.status(200).json(response);
```
Comment (leite08) [inline packages/terminology/src/router.ts:27]: nit: I'd keep the pattern we use throughout the Node/Express codebase, to `return` on these. It signifies that nothing else should be done on this code after we execute this line of code.

File: packages/terminology/src/router.ts [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+import { codeSystemImportHandler } from "./operations/codeImport";
+import { codeSystemLookupHandler } from "./operations/codeLookup";
+import { conceptMapImportHandler } from "./operations/conceptMapImport";
+import { conceptMapTranslateHandler } from "./operations/conceptMapTranslate";
+import { FhirRequest, HttpMethod } from "@medplum/fhir-router";
+import { asyncHandler } from "./util";
+
+const fhirRouter = Router();
+
+function createFhirRequest(req: Request): FhirRequest {
```
Comment (leite08) [inline packages/terminology/src/router.ts:11]: "create" gives the impression we're creating a request to be issued, while this is parsing the existing request (could also say converting from HTTP to FHIR request). Can we rename this to one of those, please?

File: packages/terminology/src/router.ts [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+    res.status(200).json(response);
+  })
+);
+
+fhirRouter.post(
+  "/CodeSystem/lookup",
+  asyncHandler(async (req: Request, res: Response) => {
+    const fhirRequest = createFhirRequest(req);
+    const response = await codeSystemLookupHandler(fhirRequest, false);
+    res.status(200).json(response);
```
Comment (leite08) [inline packages/terminology/src/router.ts:36]: We're returning arrays as the response of these endpoints. Any particular reason for this?

We standardized our responses to be a JSON object, because its easier to extend than an array.

File: packages/terminology/src/router.ts [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+  "/CodeSystem/import",
+  asyncHandler(async (req: Request, res: Response) => {
+    const fhirRequest = createFhirRequest(req);
+    const response = await codeSystemImportHandler(fhirRequest);
+    res.status(200).json(response);
+  })
+);
+
+fhirRouter.post(
+  "/CodeSystem/lookup",
```
Comment (leite08) [inline packages/terminology/src/router.ts:32]: Are we trying to implement a FHIR API of sorts, with the path using the resource in CamelCase? Our pattern AFAIK is to have those in kebab-case. If we want to have it FHIR compliant, the operation is named with `$` - [source](https://www.hl7.org/fhir/R4/codesystem-operation-lookup.html).

I'd try to simplify and just use our pattern, if being FHIR compliant is not a requirement. Its not as simple as it seems and I think it would be out of scope to be FHIR compliant.

File: LICENSE-APACHE
Comment (leite08) [reply LICENSE-APACHE:1]: What's the reasoning behind it? Like I said, the license file should be placed on the folder it belongs, as its implied that file applied to that folder - which is not the case for us.

See how GH interprets the license file that's on the root folder of the repo as the repository's license: https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository

Not to mention how confusing it would be for someone looking at our repo, two license files there (maybe more if we keep this pattern), which one applies to the root folder? Which of those licenses apply to which package/subfolder?

File: .github/workflows/pull-request_term-server.yml [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+          echo "Git ref: $(git rev-parse HEAD)"
+        shell: bash
+
+      - name: Install
+        run: |
+          pwd
+          npm install
+        working-directory: ${{env.working-directory}}
+
+      - name: Build (linters)
```
Comment (leite08) [inline .github/workflows/pull-request_term-server.yml:48]: nit: not just linting, we're using TS so this is actually compiling.

File: .github/workflows/_deploy-cdk.yml
Comment (leite08) [inline .github/workflows/_deploy-cdk.yml:1]: I think we're taking this file in the wrong direction: it's getting bulkier, slower - we should be making it leaner so it's manageable and runs within an acceptable duration. See the internal repo how the respective file deploys a single stack.

While I know its not on the scope of this PR to fix/refactor this file, I'd like to avoid adding anything here that's specific to a package.

Actually, why are we adding term server stuff here, if there's no action for it after the tests?

File: packages/infra/lib/api-stack.ts [hunk 29]
```diff
@@ -326,6 +327,18 @@ export class APIStack extends Stack {
...
+    //-------------------------------------------
+    // Terminology Server Service
+    //-------------------------------------------
+    if (!isSandbox(props.config)) {
+      createTermServerService(
+        this,
+        { ...props, generalBucket },
+        this.vpc,
+        slackNotification?.alarmAction
+      );
```
Comment (leite08) [inline packages/infra/lib/api-stack.ts:339]: We can't add more resources to the API Stack.

Any particular reason not to have this as its own stack?

File: packages/utils/package.json [hunk 30]
```diff
@@ -29,7 +29,8 @@
     "compare-resource-counts-per-file": "bash src/fhir-converter/scripts/compare_resource_counts_per_file.sh",
     "compare-total-resource-count-post-fhir-insert": "bash src/fhir-converter/scripts/compare_total_resource_count_post_fhir_insert.sh",
     "uuidv7": "ts-node src/make-uuidv7",
-    "saml-server": "ts-node src/saml/saml-server"
+    "saml-server": "ts-node src/saml/saml-server",
+    "load-test": "artillery run ./src/terminology/load-test.yml"
```
Comment (leite08) [inline packages/utils/package.json:33]: Why are we not adding this to the terminology's folder? We have testing for the FHIR Converter on the utils and I think this is not the way to go, it makes the utils package bulkier, it exposes things from the terminology to the utils (increases dependencies).

**Review by leite08 — COMMENTED** [review]
Almost all new files are using camelCase for file name, but we use kebab-case.

Also, I just realized some of the stuff came from the code import - but I can't tell what's what if it's not flagged.

I see you mentioned a couple of times that we can't have a dependency from `core`. Why is that?

File: packages/terminology/src/initTermServer.ts [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+  } else {
+    console.log(`Database file already exists at ${dbPath}`);
+  }
+
+  dbClient = createTermServerClient(dbPath);
+  await initTables(dbClient);
+  await seedCodeSystems(dbClient);
+}
+
+export function getTermServerClient(): DbClient {
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/terminology/src/initTermServer.ts:42]: This class acts as an abstraction layer or wrapper around the SQLite database. Doesn't that make this a client?

File: .github/workflows/_deploy-cdk.yml
Comment (jonahkaye) [reply .github/workflows/_deploy-cdk.yml:1]: The term server is built as part of our infra. It should pass tests during the deploy cdk no? 
I agree that its out of scope to refactor this, and I think its best to leave as is. This file is already filled with the exact same pattern of building and testing specific packages

File: LICENSE-APACHE
Comment (jonahkaye) [reply LICENSE-APACHE:1]: This is what windmill does: https://github.com/windmill-labs/windmill

which is how we got to this.

I have also seen the pattern you are describing, where the overall repo is MIT and then a sub folder has a specific license described in the readme:

https://github.com/microsoft/unilm/tree/master/layoutlmv3

Not sure which is preferred

Comment (jonahkaye) [reply LICENSE-APACHE:1]: Taking this to slack

**Review by Goncharo — APPROVED** [review]
